### ISSUE TITLE
Notable can't handle more than 10.000 MD files

### ISSUE THREAD

**Issue by suhrobina** [notable/notable#837]
- **OS Version**: Ubuntu 18.04
- **Notable Version**: 1.7.3

### Current behavior
Than you open Notable, it hangs when .notable/notes/ folder contains more than 10.000 files.

Here is bash script how I generate random MD files:
```
#!/usr/bin/env bash
for i in {0..10000}
do
  TMP_FILE=$(mktemp .notable/notes/random_XXXXXX.md)
  cat .notable/notes/test.md > $TMP_FILE
done
```
`test.md` size: 29K, q-ty of attachments: 40

### Expected behavior

For long term usage, it is important to work with a large numbers of files.

**Comment by fabiospampinato** [comment]
Absolutely, I'll make sure it handles more than 10k files. What happens on your machine exactly? Does it only take a while to open or what else?

**Comment by suhrobina** [comment]
Here is 50K file generation(Notable hangs forever):
![notable](https://user-images.githubusercontent.com/45454757/65573467-608c5f00-df84-11e9-8eb4-7ab47674278e.gif)

**Comment by fabiospampinato** [comment]
I see. Does your machine have an SSD? I'll what can be done about this, right now I'm thinking:

- We should watch for files in another process, so that the UI process stays responsive at all times.
- I've read the library I'm using for watching files (chokidar) has been updated and it now brings 50% better performance on linux.
- If you have an HDD randomly reading 50k small files will be demanding.
- Once we add a database layer between the app and the filesystem performance for these use cases should improve massively.

**Comment by suhrobina** [comment]
Ma Laptop specs: Intel(R) Core(TM) i5-4300M CPU @ 2.60GHz/

**Comment by suhrobina** [comment]
My laptop specs:

**Comment by suhrobina** [comment]
My laptop specs:
- CPU: Intel(R) Core(TM) i5-4300M CPU @ 2.60GHz;
- RAM: 12 GB SODIMM DDR3L 1600 Mhz;
- SSD: 256 GB Samsung SSD 850 Pro.

@fabiospampinato Thank You!

**Comment by fabiospampinato** [comment]
With that hardware the app must absolutely provide a much better experience, thanks for reporting the issue.

**Comment by fabiospampinato** [comment]
There's also an issue that when watching too many files we might hit the OS' limits.

This isn't an issue I can resolve in an afternoon, it would require some significant changes to how the app works today, but I think it should all be doable 👍

**Comment by fabiospampinato** [comment]
It looks like both Sublime Text and VS Code can work with 50k files, so surely Notable could be able to handle them too.

**Comment by fabiospampinato** [comment]
Some progress on this: now loading a data directory with 50k dumb notes takes ~25s rather than ~120s with just one optimization made to the app. I think we can get the app to load in somewhat reasonable times when dealing with 50k files too, it will take longer to get it to work smoothly under all scenarios with 50k notes after startup too.

Potentially after that's done I think I know how to make the app work potentially with millions of notes without making too much fuss about it, but that requires some deep architectural changes that are not going to land anytime soon. Basically it's impossible to load one million files in any reasonable time, we need to save notes into a proper database and then synchronize that asynchronously with the file system.

**Comment by fabiospampinato** [comment]
Now it takes ~19s on my machine to load 50k notes, maybe we could cut that in half optimistically with just some more optimizations, but ultimately a different database strategy would be needed to get that to load in 2 seconds or so. 

Also filesystem watching is done in another process, which improves reliability too especially when dealing with so many files.

5k notes are loaded in just under 3s on my machine, which is probably usable already.

**Comment by fabiospampinato** [comment]
Now we are down to ~9s for 50k dummy notes. Maybe we could get it down to ~5s with some more non-difficult optimizations but that sounds unrealistic.

**Comment by fabiospampinato** [comment]
We are down to ~8s now, I can't get it down much more than that without any major change, but:

- There's a PR I submitted to Electron that when merged will shave about 1s off from that.
- If notes have a metadata section, which one could make sure they have by editing them via the multi-note editor for example, then another ~2s are shaved from the current startup time because prodding the filesystem for modification date can be skipped. Those metadata sections would require parsing though which will counterbalance that somewhat.

One can get apps like VS Code to load a directory with 50k files faster than that but that's comparing apples to oranges really, VS Code is merely listing files in the sidebar, Notable is reading all of them. 

Further significant optimizations will require major changes, I'll deal with those at some point in the future, I think we could potentially bring the time required to load a 50k notes data directory down to 1s or so in some cases.
